brigade/easy.py

```diff
@@ -0,0 +1,23 @@
+from brigade.core import Brigade
+from brigade.core.configuration import Config
+from brigade.plugins.inventory.simple import SimpleInventory
+
+
+def easy_brigade(host_file="host.yaml", group_file="groups.yaml", dry_run=True, **kwargs):
```

Comment:
Would `basic_brigade` be a better name than `easy_brigade`? Easy gives the impression that the other one is hard.

Reply:
Yes, I disliked the `easy_brigade` name. 

Can we just get rid of this whole `easy.py` and have it be the `Brigade` class default:

So you can just do:

```
my_brigade = Brigade() 
```

And then in the `class Brigade` you would have:

```
     def __init__(self, inventory=None, dry_run=True, config=None, 
                       config_file="~/.brigade.yaml",
                  available_connections=None, logger=None, data=None, host_file="hosts.yaml",
                  group_file="groups.yaml"):
     ...
     if inventory is None:
         inventory=SimpleInventory(host_file, group_file),
```

I think it is more logical also to have a default config_file like `~/.brigade.yaml` and not hard-code a configuration option into easy.py.

Reply:
So the idea of `easy_brigade` (which I am fine renaming, btw) is to provide a simple "constructor" for the most common cases. We can add arguments to accommodate most use cases.

I don't like the idea of accommodating common use cases in the core API because that just adds complexity. For instance, what heppens if you pass an `inventory` and `host_file`? You ignore the latter? What if you want to also add `log_file` and `debug`? With a quick-to-access function like `easy_brigade` you have a clear demarcation of what's "business logic" and what's just "core" behavior.

> We could also move the host_file and group_file arguments to be part of Config 

That's tricky because those apply to a plugin, not to the core itself but I am tempted to agree. We could actually pass the configuration object to the inventory plugin and do something like:

```
class SimpleInventory(Inventory):
    def __init__(self, config, host_file=None, group_file=None, **kwargs):
    """
    Configuration:
            It accepts reading from the configuration the host_file and group_file
            via simple_inventory_host_file and simple_inventory_group_file
    """
    host_file = host_file or config.simple_inventory_host_file
    group_file = group_file or getattr(config, "simple_inventory_group_file")
```

Although if we are going to do that I'd like to integrate better with the `Config` object so we can also accept env variables so maybe the `Config` object should have something like:

`Config.get('simple_inventory_host_file', env_name='BRIGADE_SIMPLE_INVENTORY_HOST_FILE', default=None)`

In any case, these are two different issues:

1. Do we want `easy_brigade` (or fast_brigade, or quick_brigade or whatever other name) to accommodate common use cases with a simple import?
2. What to move to the configuration object and how to integrate better.

Reply:
For me it was mostly the name with the `easy_` prefix. I think fast_ quick_ basic would be better.

>I don't like the idea of accommodating common use cases in the core API because that just adds complexity. For instance, what heppens if you pass an inventory and host_file? You ignore the latter? What if you want to also add log_file and debug? With a quick-to-access function like easy_brigade you have a clear demarcation of what's "business logic" and what's just "core" behavior.

Completely agree with this.

Reply:
Yeah, I disagree here, but it doesn't matter much.

I am not really seeing the "business-logic" versus "core" distinction i.e. I am not seeing why some things would be categorized one way and others differently.

Two me I think we are creating two entry points into the library and the one we are categorizing as non-core will be the main one. So I think we are unnecessarily adding complexity.

Internally whether we place `easy_brigade` in the `core` directory seems irrelevant (since it will be core in practice and will cause end-users code to break if we change it improperly).

So it just seems like we are unnecessarily adding a layer of indirection for assigning default values to the class (and this causes side effects like having to document the two entry points and having to maintain more code).

But it doesn't matter much either way (i.e. it is a small amount of extra code with a pretty simple purpose).

I need to read the second half of your comments again (the comments on groups and hosts and environment variables), but probably won't have time until the weekend or next week.

Reply:
Ok, I understand where you are coming from and it was probably bad working on my side. What I mean is not that `easy_brigade` (or whatever we call it) won't be part of the "core", what I meant is that I didn't want to add business logic the `Brigade` class. I'd rather keep it agnostic and put the business logic somewhere else because that means:

a. we avoid extra complexity were not needed
b. we can provide different entry points for different use cases

As a rule of thumb I like classes and methods with very few arguments, without arguments that conflict with each other (meaning that an argument than when it has a certain value it invalidates some other argument) or with arguments that changes substantially the behavior of the method/class. I'd rather have multiple smaller methods. In summary, if it's hard to test/document, break it down into different methods/classes.

Philosophy aside, you are right pointing out we could probably do the same with default arguments/options so I am going to explore that venue.

Reply:
#73